modules/tinymce/src/core/main/ts/lookup/UserLookup.ts

```diff
@@ -0,0 +1,201 @@
+import { StructureSchema, FieldSchema } from '@ephox/boulder';
+import { Arr, Optional, Fun } from '@ephox/katamari';
+
+import Editor from '../api/Editor';
+import * as Options from '../api/Options';
+
+/**
+ * TinyMCE User Lookup API
+ * Handles user information retrieval and caching.
+ *
+ * @class tinymce.UserLookup
+ * @example
+ * // Retrieve the current user ID
+ * tinymce.activeEditor.userLookup.getCurrentUserId();
+ *
+ * // Fetch user information by IDs which returns array of promises
+ * const promises = tinymce.activeEditor.userLookup.fetchUsers(['user-1', 'user-2']);
+ * Promise.all(promises).then((users) => {
+ *   users.forEach(user => console.log('User found:', user));
+ * }).catch((error) => {
+ *   console.error('Error fetching users:', error);
+ * });
+ */
+
+type UserId = string;
+
+export interface User {
+  id: UserId;
+  name?: string;
+  avatar?: string;
+  description?: string;
+  custom?: Record<string, any>;
+}
+
+export interface UserLookup {
+  /**
+   * Retrieves the current user ID from the editor.
+   *
+   * @method getCurrentUserId
+   * @return {string} The current user ID.
+   */
+  getCurrentUserId: () => UserId;
+
+  /**
+   * Fetches user information using a provided array of userIds.
+   *
+   * @method fetchUsers
+   * @param {string[]} userIds - A list of user IDs to fetch information for.
+   * @throws {Error} Throws an error if any of the user IDs are invalid or if the fetch fails.
+   * @return {Promise<User>[]} A promise that resolves to an array of users and information about them.
+   */
+  fetchUsers: (userIds: UserId[]) => Promise<User>[];
+}
+
+const userSchema = StructureSchema.objOf([
+  FieldSchema.required('id'),
+  FieldSchema.optionString('name'),
+  FieldSchema.optionString('avatar'),
+  FieldSchema.optionString('description'),
+  FieldSchema.option('custom')
+]);
+
+const transformResult = (user: any): User => {
+  const customValue = user.custom?.getOrUndefined();
+  return {
+    id: user.id,
+    name: user.name?.getOrUndefined(),
+    avatar: user.avatar?.getOrUndefined(),
+    description: user.description?.getOrUndefined(),
```

Comment:
This is not quite the same thing. I don't know if we want properties that weren't specified to be `undefined`, they should be missing. The difference is subtle but can be important.

One benefit to your change is that it fully discards any extra properties on the returned object that we don't want. But perhaps we need to look for methods that avoid setting a property for `None` values.

Something like `Optionals.cat` but for objects 😂